File: HW2/src/container.cpp
#include <cstdio>
#include "container.h"
#include "error.h"
#include "text.h"

Container::~Container() {
    for (size_t index = 0; index < length_; ++index) {
        delete array_[index];
    }
    delete[] array_;
}

void Container::Input(FILE* file) {
    int input_type;
    long length;
    if (fscanf(file, " %zu %d ", &length, &input_type) < 2 || length < 0) {
        Error("Invalid container parameters");
    }
    length_ = length;
    array_ = new Text*[length_];
    if (input_type == 0) {
        for (size_t counter = 0; counter < length_; ++counter) {
            array_[counter] = Text::InputText(file);
        }
    } else {
        GenerateRandom(file, input_type);
    }
}

void Container::MoveFunction(FILE *file) {
    double sum = 0.0;
    for (int i = 0; i < length_; i++) {
        sum += array_[i]->Hash();
    }
    double mean = sum / length_;
    fprintf(file, "Mean: %f\n", mean);
    fprintf(file, "Move Result: \n");
    for (int i = 0; i < length_; i++) {
        if (array_[i]->Hash() <= mean) {
            array_[i]->Output(file);
        }
    }
    for (int i = 0; i < length_; i++) {
        if (array_[i]->Hash() > mean) {
            array_[i]->Output(file);
        }
    }
}

void Container::Output(FILE *file) const {
    fprintf(file, "Container has %zu elements\n", length_);
    for (size_t index = 0; index < length_; ++index) {
        fprintf(file, "%zu. ", index + 1);
        array_[index]->Output(file);
    }
}

void Container::GenerateRandom(FILE *file, int seed) {
    int cypher;
    long min_length;
    long max_length;
    long min_param;
    long max_param;
    if (fscanf(file, "%u %zu %zu %zu %zu", &cypher,
               &min_length, &max_length, &min_param, &max_param) < 5
        || cypher > 3 || min_length > max_length || min_param > max_param
        || min_length < 0 || min_param < 0) {
        Error("Invalid random generator parameters");
    }
    srand(seed);
    for (size_t counter = 0; counter < length_; ++counter) {
        array_[counter] = Text::RandomText(cypher, min_length, max_length, min_param, max_param);
    }
}

File: HW3/src/container.py
from typing import List, TextIO

from src.text import Text
from src.text_factory import input_text, random_text


class Container:
    def __init__(self, array: List[Text]) -> None:
        self.array = array

    @classmethod
    def from_file(cls, input_file: TextIO):
        try:
            length = int(input_file.readline())
            if length < 0:
                raise
        except Exception:
            raise ValueError("Invalid container parameters")
        return cls([input_text(input_file) for _ in range(length)])

    @classmethod
    def from_random_elements(cls, length: int):
        return cls([random_text() for _ in range(length)])

    def __len__(self) -> int:
        return len(self.array)

    def move_function(self, output_file: TextIO):
        sum = 0.0
        for elem in self.array:
            sum += elem.hash
        if len(self) != 0:
            mean = sum / len(self)
        else:
            mean = 0
        print(f"Mean: {mean}", file=output_file)
        print("Move Result: ", file=output_file)
        for elem in self.array:
            if elem.hash <= mean:
                elem.output(output_file)
        for elem in self.array:
            if elem.hash > mean:
                elem.output(output_file)

    def output(self, output_file: TextIO) -> None:
        print("Container has", len(self), "elements", file=output_file)
        for index in range(len(self)):
            print(index + 1, end=". ", file=output_file)
            self.array[index].output(output_file)
